src/main/java/com/wojciechkocik/usage/service/TimeSpentCrossMidnightServiceImpl.java

```diff
@@ -0,0 +1,77 @@
+package com.wojciechkocik.usage.service;
+
+import com.wojciechkocik.usage.dto.DailyUsage;
+import org.springframework.stereotype.Service;
+
+import java.time.Duration;
+import java.time.LocalTime;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Implementation of {@link TimeSpentCrossMidnightService} using java 8 features
+ *
+ * @author Wojciech Kocik
+ * @since 24.03.2017
+ */
+@Service
+public class TimeSpentCrossMidnightServiceImpl implements TimeSpentCrossMidnightService {
+
+    @Override
+    public List<DailyUsage> divideDaysWithTimeSessionCrossedMidnight(List<DailyUsage> dailyUsagesForCourse) {
+
+        List<DailyUsage> dividedDailyUsages = new ArrayList<>();
+
+        dailyUsagesForCourse.forEach(dailyUsage -> {
+            ZonedDateTime timeDateStarted = dailyUsage.getDateTime();
+
+            long sessionTime = dailyUsage.getTime();
+
+            long epochBeforeSession = timeDateStarted.toEpochSecond();
+            long epochAfterSession = timeDateStarted.plusSeconds(sessionTime).toEpochSecond();
+            long epochDiff = epochAfterSession - epochBeforeSession;
+
+            long untilMidnight = timeDateStarted.until(getMidnightOfDate(timeDateStarted.plusDays(1)), ChronoUnit.SECONDS);
+            long restEpoch = epochDiff - untilMidnight;
+
+            if(sessionTime <= 0){
+                return;
```

Comment:
If you have to check such conditions inside your business logic, than either:
1. You assumes, that it might apear an error in other module in you system, so you double checks that sessionTime > 0. But in such case, I would at least used log.error() with appropriate message.
This assumption also implies, that the tests of other parts of the system are weak.
2. Negative or 0 sessionTime is for some reason OK and controller layer does not reject such values, but your business logic here assumes such values as invalid, thus make this validation. In such case this check would be probably OK, but I would have problem to understand, what a negative sessionTime duration actually means.